index.js

```diff
@@ -1,109 +1,145 @@
-const outputElmFilename = path.join(outputDir, `${runModuleName}.elm`);
-const outputCompiledFilename = path.join(outputDir, 'run-elm-main.js');
-
-function cleanup() {
-  return Promise.all([
-    unlink(outputElmFilename).catch(() => {}),
-    unlink(outputCompiledFilename).catch(() => {}),
-  ]);
-}
-
-function getTemplate() {
-  return readFile(templatePath, 'utf-8');
-}
-
-function createElmFile(template) {
-  const contents = template.replace('{module}', moduleName);
-  return writeFile(outputElmFilename, contents, 'utf-8');
-}
-
-function compileElmFile() {
-  sh.cd(outputDir);
+    } catch (err) {
+      throw new Error(`File '${userModulePath}' does not exist`);
+    }
+    if (!userModulePath.match(/\.elm$/i, '')) {
+      throw new Error(`File '${userModulePath}' should have .elm file extension`);
+    }
 
-  return compileToString([outputElmFilename], { yes: true })
-    .then(contents => writeFile(outputCompiledFilename, contents));
-}
+    // ensure project folder is adequate
+    try {
+      const projectDirStats = await stat(projectDir);
+      if (!projectDirStats.isDirectory()) {
+        throw new Error();
+      }
+    } catch (err) {
+      throw new Error(`Provided --project-dir '${rawProjectDir}' is not a directory`);
+    }
+    if (userModulePath.indexOf(`${projectDir}/`) !== 0) {
+      throw new Error(
+        `File \`${userModulePath}\` must be located within --project-dir \`${rawProjectDir}\``
+      );
+    }
 
-function runElmFile() {
-  return new Promise((resolve) => {
-    const { worker } = require(outputCompiledFilename)[runModuleName];
+    // read RunElmMain template
+    await sh.cd(projectDir);
+    let template;
+    try {
+      template = await readFile(templatePath, 'utf-8');
+    } catch (err) {
+      throw new Error(`Elm file '${templatePath}' does not exist`);
+    }
 
-    worker().ports.sendOutput.subscribe((output) => {
-      console.log(output);
-      resolve();
+    // create RunELmMain file from template
+    const mainModuleContents = template
+      .replace('{mainModule}', mainModule)
+      .replace('{userModule}', userModule)
+      .replace('{output}', outputVarName);
+    await writeFile(mainModuleFilename, mainModuleContents, 'utf-8');
+
+    // compile main module
+    const contents = await compileToString([mainModuleFilename], { yes: true });
+    await writeFile(outputCompiledFilename, contents);
+
+    // run compiled elm file
+    await new Promise((resolve) => {
+      const { worker } = require(outputCompiledFilename)[mainModule];
+      worker().ports.sendOutput.subscribe((output) => {
+        console.log(output);
+        resolve();
+      });
     });
-  });
-}
-
-function handleError(err) {
-  let message;
-
-  if (typeof err === 'object' && 'message' in err) {
-    if (/does not expose `output`/.test(err.message)) {
-      message = `Elm module \`${moduleName}\` does not define a String \`output\` variable.`;
+  } catch (err) {
```

Comment:
This catch block represents all possible failures and thus standardizes error printing.